### PR TITLE
fix(web): corrects Android over-deletion of selected text, other context diffs involving selected text

### LINKED ISSUE #10645: bug(android): typing a character when text is selected deletes past the end of the selection
Keyman 17.0.256-alpha on Android, Samsung A53/5G phone, Android 14

If you make a text selection (all apps that I've tried), and then type a character, Keyman deletes the selection, and then continues past and deletes too much.

https://github.com/keymanapp/keyman/assets/4498365/11129486-5a8f-48d1-a140-3126d27da29a

### PR DESCRIPTION
**Pull request by jahorton**
Fixes #10645.

Back in 12.0, when the `Transcription` and `Transform` abstractions were developed to facilitate predictive text, we didn't have much robustness in the way of handling cases where text was selected.  We've added that in many other places within Web since... but the construction of `Transform`s as the difference between two context states never got updated accordingly.  At this point in time, this results in selected-text deletion being treated as right-deleted text.

The _correct_ behavior would be to treat the selected-text as invisible, implicitly removed whenever a transform is applied... and thus not a right-deletion.  The Android-side code is patterned after this expectation already; refer to https://github.com/keymanapp/keyman/issues/10645#issuecomment-1931098254.

## User Testing

Fortunately, we have a lot of unit tests in place already to verify that `Transform` construction is still correct.  That said...

**TEST_ANDROID_REPLACE_SELECTION**:  Using the Keyman for Android app...

1. Type the following text:  `This is a test and some more text here.`
2. Select `test and`.
3. Type `B`.
4. Verify that the end result is `This is a B some more text here.`

### CODE REVIEW

File: web/src/engine/main/src/keymanEngine.ts [hunk 1]
```diff
@@ -196,6 +196,9 @@ export default class KeymanEngine<
 
     config.initialize(optionSpec);
 
+    // Initialize supplementary plane string extensions
+    String.kmwEnableSupplementaryPlane(true);
```
Comment (mcdurdin) [inline web/src/engine/main/src/keymanEngine.ts:200]: This turns on non-BMP string processing in all our app/targets -- not just web browsers, but also when embedded in iOS and Android apps.

File: common/web/keyboard-processor/src/text/stringDivergence.ts [hunk 2]
```diff
@@ -0,0 +1,68 @@
...
+  for(; commonLeft ? index <= end: index >= end; index += inc) {
+    if(str1.charAt(index) != str2.charAt(index + offset)) {
+      break;
+    }
+  }
+
+  // `index` corresponds to the first char that is different _in the direction indicated by inc_.
+
+  // if commonLeft, high surrogate; if commonRight, low surrogate.
+  const commonPotentialSurrogate = str1.charCodeAt(index - inc);
```
Comment (mcdurdin) [inline common/web/keyboard-processor/src/text/stringDivergence.ts:46]: This will fail if the first char in the string doesn't match because then it will be passed `charCodeAt(-1)`

File: common/web/keyboard-processor/tests/node/transcriptions.js [hunk 3]
```diff
@@ -1,20 +1,129 @@
...
-describe("Transcriptions and Transforms", function() {
-  var toSupplementaryPairString = function(code){
-    var H = Math.floor((code - 0x10000) / 0x400) + 0xD800;
-    var L = (code - 0x10000) % 0x400 + 0xDC00;
+const toSupplementaryPairString = function(code){
+  var H = Math.floor((code - 0x10000) / 0x400) + 0xD800;
+  var L = (code - 0x10000) % 0x400 + 0xDC00;
+
+  return String.fromCharCode(H, L);
+}
```
Comment (mcdurdin) [inline common/web/keyboard-processor/tests/node/transcriptions.js:15]: String.fromCodepoint is available to us from Chrome 41. That's good enough for unit tests.

File: common/web/keyboard-processor/src/text/stringDivergence.ts [hunk 4]
```diff
@@ -0,0 +1,68 @@
+/**
+ * Returns the index for the code point divergence point in code unit coordinates.
+ * @param str1
+ * @param str2
+ * @param commonRight If false or undefined, asserts a common prefix to the strings.  If true, asserts a common suffix.
+ * @returns The code unit indices within each string for the start of the code point not common to both.
+ */
+export function searchStringDivergence(str1: string, str2: string, commonRight?: boolean): [number, number] {
+  let maxInterval = Math.min(str1.length, str2.length) - 1;
+  const commonLeft = !commonRight;
```
[the remark this answers is not in the document]
Comment (jahorton) [reply common/web/keyboard-processor/src/text/stringDivergence.ts:10]: I did that at first, but I think the natural default is `commonLeft`.  `false` and `undefined` are both truthy, so... I went this way.

File: common/web/keyboard-processor/src/text/stringDivergence.ts [hunk 5]
```diff
@@ -0,0 +1,68 @@
+/**
+ * Returns the index for the code point divergence point in code unit coordinates.
+ * @param str1
+ * @param str2
+ * @param commonRight If false or undefined, asserts a common prefix to the strings.  If true, asserts a common suffix.
```
[the remark this answers is not in the document]
Comment (jahorton) [reply common/web/keyboard-processor/src/text/stringDivergence.ts:5]: > And instead of `commonRight` we should say `commonSuffix` to avoid confusion with RTL/LTR scripts.

Fair.

> I think it will be better split.

I find myself disagreeing for a few reasons:
- While the "common prefix" version is fairly intuitive... the "common suffix" one is _markedly_ less so, or at least it is to me.  I find that having the merged method _heavily_ clarifies the logic underlying the "common suffix" pathway.
- Any future fixes, like with https://github.com/keymanapp/keyman/pull/10662#discussion_r1480981258 would be subjected to WETness, so it'd easily be possible to miss fixing one version or the other.  For example, the linked comment was directed toward the "prefix" variant, but the "suffix" variant also has an equivalent case.
- Being WET = more filesize cost.  Not _that_ big a footprint here... I think... but there is still a difference.

[the remark this answers is not in the document]
Comment (mcdurdin) [reply common/web/keyboard-processor/src/text/stringDivergence.ts]: Complexity argument:
* The original function includes 5 separate branches based on that parameter -- including one buried in the test predicate of the `for` loop
* `calculateCommonPrefix` has 2 variables where the original function has 12. `calculateCommonSuffix` should have 3 variables.

Yes, it's a little WETter, but it's far more maintainable and easier to reason against. There are no performance benefits to the DRYer solution -- it will necessarily be slower with the extra tests. The size benefit is marginal, if any.

[the remark this answers is not in the document]
Comment (jahorton) [reply common/web/keyboard-processor/src/text/stringDivergence.ts]: > * `calculateCommonPrefix` has 2 variables where the original function has 12. `calculateCommonSuffix` should have 3 variables.

I now count (effectively) 8.  This does _not_ include:
- commonPotentialSurrogate
- divergentChar1
- divergentChar2

Those three can easily be eliminated by having their assigned values inlined for their single usage within the method.  It's not fair to count them against the original implementation.

- isHigh
- isLow

Those two aren't included, as you didn't count them for the split versions.

I could eliminate more variables for the merged version, but the result would almost certainly be _less_ clear if I did so.  I find breaking things down into pieces with them helps with clarity here.

**Comment by bharanidharanj** [comment]
# Test Results

* **TEST_ANDROID_REPLACE_SELECTION (PASSED):** Tested with the attached PR build (ver 17.0.258-alpha-test-10662) in the Android 13 Mobile and here is my observation: 1. Typed the given text. 2. Selected and edited the mentioned text. 3. Verified that the result is 'This is a B some more text here' shown on the screen.

File: common/web/keyboard-processor/src/text/stringDivergence.ts [hunk 6]
```diff
@@ -0,0 +1,88 @@
...
+    const isHigh = (charCode: number) => charCode >= 0xD800 && charCode <= 0xDBFF;
+    const isLow =  (charCode: number) => charCode >= 0xDC00 && charCode <= 0xDFFF;
+    const commonChecker = commonSuffix ? isLow : isHigh;
+    const divergentChecker = commonSuffix ? isHigh : isLow;
+
+    // If the last common char qualifies as a direction-appropriate SMP surrogate...
+    if(commonChecker(commonPotentialSurrogate)) {
+      // And one of the two divergent chars is a qualifying match - a surrogate
+      // of the opposite type...
+      if(divergentChecker(divergentChar1) || divergentChecker(divergentChar2)) {
```
Comment (mcdurdin) [inline common/web/keyboard-processor/src/text/stringDivergence.ts:79]: I am wondering if we really need to test the next code unit at all? If the surrogate is isolated then it's already going to be a problem...

And that makes me wonder about split surrogates at start or end of string, what happens there?

File: common/web/keyboard-processor/src/text/stringDivergence.ts [hunk 7]
```diff
@@ -0,0 +1,88 @@
...
+    if(str1.charAt(index) != str2.charAt(index + offset)) {
+      break;
+    }
+  }
+
+  // Step 2:  Ensure that we're not splitting a surrogate pair.
+
+  // `index` corresponds to the first char that is different _in the direction indicated by inc_.
+  // If it's the start position, it can't split a (completed) surrogate pair.
+  if(index != start) {
```
Comment (mcdurdin) [inline common/web/keyboard-processor/src/text/stringDivergence.ts:63]:
```suggestion
  if(index != start && index != end) {
```

I think this test is needed because `str1.charCodeAt(end)` would be off the end of the string.

File: common/web/keyboard-processor/src/text/stringDivergence.ts [hunk 8]
```diff
@@ -0,0 +1,88 @@
...
+    // if commonLeft, high surrogate; if commonRight, low surrogate.
+    const commonPotentialSurrogate = str1.charCodeAt(index - inc);
+    // Opposite surrogate type from the previous variable.
+    const divergentChar1 = str1.charCodeAt(index);
+    const divergentChar2 = str2.charCodeAt(index + offset);
+
+    const isHigh = (charCode: number) => charCode >= 0xD800 && charCode <= 0xDBFF;
+    const isLow =  (charCode: number) => charCode >= 0xDC00 && charCode <= 0xDFFF;
+    const commonChecker = commonSuffix ? isLow : isHigh;
+    const divergentChecker = commonSuffix ? isHigh : isLow;
```
Comment (mcdurdin) [inline common/web/keyboard-processor/src/text/stringDivergence.ts:73]: I'm not a fan of these aliases -- they make it harder to understand what is actually being tested. I don't really like `isStartOfSurrogatePair` because of the directionality issues but it may be somewhat clearer?

```suggestion
    const isStartOfSurrogatePair = commonSuffix ? isLow : isHigh;
    const isEndOfSurrogatePair = commonSuffix ? isHigh : isLow;
```

File: common/web/keyboard-processor/src/text/stringDivergence.ts [hunk 9]
```diff
@@ -0,0 +1,88 @@
...
+
+    // If the last common char qualifies as a direction-appropriate SMP surrogate...
```
Comment (mcdurdin) [inline common/web/keyboard-processor/src/text/stringDivergence.ts:75]:
```suggestion
    // If the last common char qualifies as a direction-appropriate start of the surrogate pair...
```

File: common/web/keyboard-processor/tests/node/transcriptions.js [hunk 10]
```diff
@@ -1,22 +1,161 @@
...
+
+    it("BMP edge cases", () => {
+      const result1 = searchStringDivergence("applesauce", "applesauce", false);
+      assert.equal(result1, 10);
+
+      const result2 = searchStringDivergence("applesauce", "banana bread", false);
+      assert.equal(result2, 0);
+    });
+
+    it("SMP text", () => {
```
Comment (mcdurdin) [inline common/web/keyboard-processor/tests/node/transcriptions.js:42]:
```suggestion
    it("non-BMP text", () => {
```

File: common/web/keyboard-processor/tests/node/transcriptions.js [hunk 11]
```diff
@@ -1,22 +1,161 @@
...
+      const smp_ify = (str) => str.split('').map(ss).join('');
+
+      const result1 = searchStringDivergence(
+        smp_ify('apple'),
+        smp_ify('applause'),
+        false
+      );
+
+      // 2 per SMP char; is in code-unit... units.
```
Comment (mcdurdin) [inline common/web/keyboard-processor/tests/node/transcriptions.js:51]:
```suggestion
      // 2 per non-BMP char; is in code-unit... units.
```

File: common/web/keyboard-processor/tests/node/transcriptions.js [hunk 12]
```diff
@@ -1,22 +1,161 @@
...
+      // If the two are equal...
+      const result1 = searchStringDivergence("post-caret text", "post-caret text", true);
+      assert.equal(result1, -1);
+
+      // If the two are completely different...
+      const result2 = searchStringDivergence("post-caret text", "supercalifragilistic", true);
+      assert.equal(result2, "post-caret text".length-1);
+    })
+
+    it("SMP text", () => {
```
Comment (mcdurdin) [inline common/web/keyboard-processor/tests/node/transcriptions.js:109]:
```suggestion
    it("non-BMP text", () => {
```

File: common/web/keyboard-processor/tests/node/transcriptions.js [hunk 13]
```diff
@@ -1,22 +1,161 @@
...
+
+      //   att|endance
+      // trans|endance
+      const result1 = searchStringDivergence(
+        smp_ify("attendance"),
+        smp_ify("transcendance"),
+        true
+      );
+
+      // 2 per SMP char; is in code-unit... units.
```
Comment (mcdurdin) [inline common/web/keyboard-processor/tests/node/transcriptions.js:120]:
```suggestion
      // 2 per non-BMP char; is in code-unit... units.
```

File: common/web/keyboard-processor/tests/node/transcriptions.js [hunk 14]
```diff
@@ -1,22 +1,161 @@
...
+      const result2 = searchStringDivergence(
+        smp_ify("transcendance"),
+        smp_ify("happenstance"),
+        true
+      );
+      assert.equal(result2, 17);
+
+    });
 
+    it("SMP edge cases", () => {
```
Comment (mcdurdin) [inline common/web/keyboard-processor/tests/node/transcriptions.js:135]:
```suggestion
    it("non-BMP edge cases", () => {
```

File: common/web/keyboard-processor/src/text/stringDivergence.ts [hunk 15]
```diff
@@ -0,0 +1,88 @@
+/**
+ * Returns the index for the code point divergence point in code unit coordinates.
+ * @param str1
+ * @param str2
+ * @param commonSuffix If false, asserts a common prefix to the strings.  If true, asserts a common suffix.
+ * @returns The code unit index within `str1` for the start of the code point not common to both.
+ */
+export function searchStringDivergence(str1: string, str2: string, commonSuffix: boolean): number {
+  /**
```
[the remark this answers is not in the document]
Comment (jahorton) [reply common/web/keyboard-processor/src/text/stringDivergence.ts:9]: image

Is it so bad to have Intellisense-friendly documentation on local variables?

[the remark this answers is not in the document]
Comment (mcdurdin) [reply common/web/keyboard-processor/src/text/stringDivergence.ts:9]: Fine. But is the function so complex that you need intellisense docs on your local variables?!

File: common/web/keyboard-processor/src/text/stringDivergence.ts [hunk 16]
```diff
@@ -0,0 +1,88 @@
...
+  // If it's the start position, it can't split a (completed) surrogate pair.
+  if(index != start) {
+    // if commonLeft, high surrogate; if commonRight, low surrogate.
+    const commonPotentialSurrogate = str1.charCodeAt(index - inc);
+    // Opposite surrogate type from the previous variable.
+    const divergentChar1 = str1.charCodeAt(index);
+    const divergentChar2 = str2.charCodeAt(index + offset);
+
+    const isHigh = (charCode: number) => charCode >= 0xD800 && charCode <= 0xDBFF;
+    const isLow =  (charCode: number) => charCode >= 0xDC00 && charCode <= 0xDFFF;
```
[the remark this answers is not in the document]
Comment (jahorton) [reply common/web/keyboard-processor/src/text/stringDivergence.ts:71]: I'm not opposed.  Part of me was thinking about moving them to `common/web/utils`.  Would that be reasonable, or would `common/web/types` be better?

[the remark this answers is not in the document]
Comment (mcdurdin) [reply common/web/keyboard-processor/src/text/stringDivergence.ts:71]: common/web/types/src/util/util.ts already has:

https://github.com/keymanapp/keyman/blob/107e8e98f94d955984059440c27e1597ae241001/common/web/types/src/util/util.ts#L241-L254

So these just need to be exported and then they'll be available as `util.Uni_IsSurrogate1` etc

File: common/web/keyboard-processor/src/text/stringDivergence.ts (hunk 6 above)
Comment (jahorton) [reply common/web/keyboard-processor/src/text/stringDivergence.ts:79]: > And that makes me wonder about split surrogates at start or end of string, what happens there?

Like... if the caret itself is splitting a surrogate?  Or just if the surrogate pair is the first thing examined?

Comment (mcdurdin) [reply common/web/keyboard-processor/src/text/stringDivergence.ts:79]: e.g. pasted into a text field, a string with either starts or ends with a split surrogate

File: common/web/keyboard-processor/src/text/stringDivergence.ts (hunk 7 above)
Comment (jahorton) [reply common/web/keyboard-processor/src/text/stringDivergence.ts:63]: It's not bad to have for clarity, to be sure, and it's an easy add.  Sure.

It _technically_ was fine as it was:

```
> ''.charCodeAt(1000)
> NaN
```

From there, `NaN <= value` or `NaN >= value` always returns `false`... and so the implementation of the surrogate-checking functions happened to leverage this detail.  It's why the related unit tests were already passing.  

I'm not going to claim use of these `NaN` properties was intentional, though.  And, of course, their use isn't obvious from the code, so adding something that's more explicit is reasonable.

File: common/web/keyboard-processor/src/text/stringDivergence.ts (hunk 6 above)
Comment (jahorton) [reply common/web/keyboard-processor/src/text/stringDivergence.ts:79]: > I am wondering if we really need to test the next code unit at all? If the surrogate is isolated then it's already going to be a problem...

This aspect of the logic is preserved from the original implementation.  I'm not saying that your question here is invalid... just stating the reason it's not been addressed so far within this PR.

File: common/web/keyboard-processor/src/text/stringDivergence.ts (hunk 7 above)
Comment (mcdurdin) [reply common/web/keyboard-processor/src/text/stringDivergence.ts:63]: Cool cool. `ba'NaN'a` sounding good right now.

File: common/web/keyboard-processor/src/text/stringDivergence.ts (hunk 6 above)
Comment (mcdurdin) [reply common/web/keyboard-processor/src/text/stringDivergence.ts:79]: OK let's continue with original logic because it has not been a problem so far as we can tell.

File: common/web/keyboard-processor/src/text/stringDivergence.ts (hunk 16 above)
[the remark this answers is not in the document]
Comment (jahorton) [reply common/web/keyboard-processor/src/text/stringDivergence.ts:71]: Looks like we may have to wait until ES5 is out of the picture to use the `common/web/types` version; `esbuild` refuses to downcompile for our ES5 target.  We can get away with it at present for _types_, but any actual functional code references lead to bundling errors as a result.

I'll put a version of this in `common/web/utils` for now, with a note to merge it in Keyman 18 / once we drop the ES5 target.

File: common/web/keyboard-processor/src/text/stringDivergence.ts [hunk 17]
```diff
@@ -0,0 +1,88 @@
...
+  // Step 2:  Ensure that we're not splitting a surrogate pair.
+
+  // `index` corresponds to the first char that is different _in the direction indicated by inc_.
+  // If it's the start position, it can't split a (completed) surrogate pair.
+  if(index != start) {
+    // if commonLeft, high surrogate; if commonRight, low surrogate.
+    const commonPotentialSurrogate = str1.charCodeAt(index - inc);
+    // Opposite surrogate type from the previous variable.
+    const divergentChar1 = str1.charCodeAt(index);
+    const divergentChar2 = str2.charCodeAt(index + offset);
```
[the remark this answers is not in the document]
Comment (jahorton) [reply common/web/keyboard-processor/src/text/stringDivergence.ts:72]: Resolved with the suggested fix to https://github.com/keymanapp/keyman/pull/10662#discussion_r1482327105.

**Comment by keyman-server** [comment]
Changes in this pull request will be available for download in [Keyman version 17.0.269-beta](https://keyman.com/downloads/releases/beta/17.0.269)
